File: android/src/main/kotlin/com/patricktailor/snowplow_flutter_tracker/util/EventUtil.kt
package com.patricktailor.snowplow_flutter_tracker.util

import com.snowplowanalytics.snowplow.tracker.events.*
import com.snowplowanalytics.snowplow.tracker.payload.SelfDescribingJson

class EventUtil {
    companion object {
        fun getSelfDescribingEvent(json: Map<String, Any>?): SelfDescribing {
            val selfDescribingJson = getSelfDescribingJson(json)
            return SelfDescribing.builder().eventData(selfDescribingJson).build()
        }

        fun getStructuredEvent(json: Map<String, Any>?): Structured {
            val builder = Structured.builder()

            if (json?.get("category") != null) {
                builder.category(json["category"] as String)
            }
            if (json?.get("action") != null) {
                builder.action(json["action"] as String)
            }
            if (json?.get("label") != null) {
                builder.label(json["label"] as String)
            }
            if (json?.get("property") != null) {
                builder.property(json["property"] as String)
            }
            if (json?.get("value") != null) {
                builder.value(json["value"] as Double)
            }

            return builder.build()
        }

        fun getScreenViewEvent(json: Map<String, Any>?): ScreenView {
            val builder = ScreenView.builder()

            if (json?.get("name") != null) {
                builder.name(json["name"] as String)
            }
            if (json?.get("type") != null) {
                builder.type(json["type"] as String)
            }
            if (json?.get("transitionType") != null) {
                builder.transitionType(json["transitionType"] as String)
            }
            if (json?.get("previousName") != null) {
                builder.previousName(json["previousName"] as String)
            }
            if (json?.get("previousType") != null) {
                builder.previousType(json["previousType"] as String)
            }

            return builder.build()
        }

        fun getPageViewEvent(json: Map<String, Any>?): PageView {
            val builder = PageView.builder()
            
            if (json?.get("pageUrl") != null) {
                builder.pageUrl(json["pageUrl"] as String)
            }
            if (json?.get("pageTitle") != null) {
                builder.pageTitle(json["pageTitle"] as String)
            }
            if (json?.get("referrer") != null) {
                builder.referrer(json["referrer"] as String)
            }
            
            return builder.build()
        }

        fun getTimingEvent(json: Map<String, Any>?): Timing {
            val builder = Timing.builder()

            if (json?.get("category") != null) {
                builder.category(json["category"] as String)
            }
            if (json?.get("variable") != null) {
                builder.variable(json["variable"] as String)
            }
            if (json?.get("timing") != null) {
                builder.timing(json["timing"] as Int)
            }
            if (json?.get("label") != null) {
                builder.label(json["label"] as String)
            }

            return builder.build()
        }

        fun getEcommerceTransactionEvent(json: Map<String, Any>?): EcommerceTransaction {
            val builder = EcommerceTransaction.builder()

            if (json?.get("orderId") != null) {
                builder.orderId(json["orderId"] as String)
            }
            if (json?.get("totalValue") != null) {
                builder.totalValue(json["totalValue"] as Double)
            }
            if (json?.get("affiliation") != null) {
                builder.affiliation(json["affiliation"] as String)
            }
            if (json?.get("taxValue") != null) {
                builder.taxValue(json["taxValue"] as Double)
            }
            if (json?.get("shipping") != null) {
                builder.shipping(json["shipping"] as Double)
            }
            if (json?.get("city") != null) {
                builder.city(json["city"] as String)
            }
            if (json?.get("state") != null) {
                builder.state(json["state"] as String)
            }
            if (json?.get("country") != null) {
                builder.country(json["country"] as String)
            }
            if (json?.get("currency") != null) {
                builder.currency(json["currency"] as String)
            }
            if (json?.get("items") != null) {
                val items = (json["items"] as List<Map<String, Any>>).map(::getEcommerceTransactionItem)
                builder.items(items)
            }

            return builder.build()
        }

        fun getConsentGrantedEvent(json: Map<String, Any>?): ConsentGranted {
            val builder = ConsentGranted.builder()

            if (json?.get("expiry") != null) {
                builder.expiry(json["expiry"] as String)
            }
            if (json?.get("documentId") != null) {
                builder.documentId(json["documentId"] as String)
            }
            if (json?.get("documentVersion") != null) {
                builder.documentVersion(json["documentVersion"] as String)
            }
            if (json?.get("documentName") != null) {
                builder.documentName(json["documentName"] as String)
            }
            if (json?.get("documentDescription") != null) {
                builder.documentDescription(json["documentDescription"] as String)
            }
            if (json?.get("consentDocuments") != null) {
                val consentDocuments = (json["consentDocuments"] as List<Map<String, Any>>).map(::getConsentDocuments)
                builder.consentDocuments(consentDocuments)
            }

            return builder.build()
        }

        fun getConsentWithdrawnEvent(json: Map<String, Any>?): ConsentWithdrawn {
            val builder = ConsentWithdrawn.builder()
            
            if (json?.get("all") != null) {
                builder.all(json["all"] as Boolean)
            }
            if (json?.get("documentId") != null) {
                builder.documentId(json["documentId"] as String)
            }
            if (json?.get("documentVersion") != null) {
                builder.documentVersion(json["documentVersion"] as String)
            }
            if (json?.get("documentName") != null) {
                builder.documentName(json["documentName"] as String)
            }
            if (json?.get("documentDescription") != null) {
                builder.documentDescription(json["documentDescription"] as String)
            }
            if (json?.get("consentDocuments") != null) {
                val consentDocuments = (json["consentDocuments"] as List<Map<String, Any>>).map(::getConsentDocuments)
                builder.consentDocuments(consentDocuments)
            }
            
            return builder.build()
        }

        private fun getSelfDescribingJson(json: Map<String, Any>?): SelfDescribingJson {
            val eventData = json?.get("eventData") as Map<String, Any>?
            return SelfDescribingJson(eventData?.get("schema") as String, eventData["payload"] as Map<String, Any>)
        }

        private fun getEcommerceTransactionItem(json: Map<String, Any>?): EcommerceTransactionItem {
            val builder = EcommerceTransactionItem.builder()

            if (json?.get("itemId") != null) {
                builder.itemId(json["itemId"] as String)
            }
            if (json?.get("sku") != null) {
                builder.sku(json["sku"] as String)
            }
            if (json?.get("price") != null) {
                builder.price(json["price"] as Double)
            }
            if (json?.get("quantity") != null) {
                builder.quantity(json["quantity"] as Int)
            }
            if (json?.get("name") != null) {
                builder.name(json["name"] as String)
            }
            if (json?.get("category") != null) {
                builder.category(json["category"] as String)
            }
            if (json?.get("currency") != null) {
                builder.currency(json["currency"] as String)
            }

            return builder.build()
        }

        private fun getConsentDocuments(json: Map<String, Any>?): ConsentDocument {
            val builder = ConsentDocument.builder()

            if (json?.get("documentId") != null) {
                builder.documentId(json["documentId"] as String)
            }
            if (json?.get("documentVersion") != null) {
                builder.documentVersion(json["documentVersion"] as String)
            }
            if (json?.get("documentName") != null) {
                builder.documentName(json["documentName"] as String)
            }
            if (json?.get("documentDescription") != null) {
                builder.documentDescription(json["documentDescription"] as String)
            }

            return builder.build()
        }
    }
}

File: ios/Classes/EventUtil.swift
//
//  EventUtil.swift
//  snowplow_flutter_tracker
//
//  Created by Szabó Patrik on 2020. 06. 30..
//

import Foundation
import SnowplowTracker

struct EventUtil {
    static public func getPageView(dict: [String: Any]?) -> SPPageView {
        return SPPageView.build { (SPPageViewBuilder) in
            SPPageViewBuilder.setPageUrl(dict?["pageUrl"] as! String)
            SPPageViewBuilder.setPageTitle(dict?["pageTitle"] as? String)
            SPPageViewBuilder.setReferrer(dict?["referrer"] as? String)
            SPPageViewBuilder.setContexts(
                getContexts(dict?["contexts"] as? [[String: Any]])
            )
        }
    }
    
    static public func getStructured(dict: [String: Any]?) -> SPStructured {
        return SPStructured.build { (SPStructuredBuilder) in
            SPStructuredBuilder.setCategory(dict?["category"] as! String)
            SPStructuredBuilder.setAction(dict?["action"] as! String)

            SPStructuredBuilder.setLabel(dict?["label"] as? String)
            SPStructuredBuilder.setProperty(dict?["property"] as? String)
            (dict?["value"] as? Double).map(SPStructuredBuilder.setValue)

            SPStructuredBuilder.setContexts(
                getContexts(dict?["contexts"] as? [[String: Any]])
            )
        }
    }
    
    static public func getUnstructured(dict: [String: Any]?) -> SPUnstructured {
        let eventData: SPSelfDescribingJson = getSelfDescribingJson(dict: dict?["eventData"] as! [String: Any])
        
        return SPUnstructured.build { (SPUnstructuredBuilder) in
            SPUnstructuredBuilder.setEventData(eventData)
            SPUnstructuredBuilder.setContexts(
                getContexts(dict?["contexts"] as? [[String: Any]])
            )
        }
    }
    
    static public func getScreenView(dict: [String: Any]?) -> SPScreenView {
        return SPScreenView.build { (SPScreenViewBuilder) in
            SPScreenViewBuilder.setName(dict?["name"] as! String)
            SPScreenViewBuilder.setType(dict?["type"] as? String)
            SPScreenViewBuilder.setTransitionType(dict?["transitionType"] as? String)
            SPScreenViewBuilder.setPreviousScreenName(dict?["previousName"] as? String)
            SPScreenViewBuilder.setPreviousScreenType(dict?["previousType"] as? String)
            SPScreenViewBuilder.setContexts(
                getContexts(dict?["contexts"] as? [[String: Any]])
            )
        }
    }
    
    static public func getConsentWithdrawn(dict: [String: Any]?) -> SPConsentWithdrawn {
        let documents = getConsentDocuments(payloads: (dict?["consentDocuments"] as! [[String: Any]]))
        
        return SPConsentWithdrawn.build { (SPConsentWithdrawnBuilder) in
            (dict?["all"] as? Bool).map(SPConsentWithdrawnBuilder.setAll)
            (dict?["documentId"] as? String).map(SPConsentWithdrawnBuilder.setDocumentId)
            (dict?["documentVersion"] as? String).map(SPConsentWithdrawnBuilder.setVersion)
            (dict?["documentName"] as? String).map(SPConsentWithdrawnBuilder.setName)
            (dict?["documentDescription"] as? String).map(SPConsentWithdrawnBuilder.setDescription)

            SPConsentWithdrawnBuilder.setDocuments(documents)
            SPConsentWithdrawnBuilder.setContexts(
                getContexts(dict?["contexts"] as? [[String: Any]])
            )
        }
    }
    
    static public func getConsentGranted(dict: [String: Any]?) -> SPConsentGranted {
        let documents = getConsentDocuments(payloads: (dict?["consentDocuments"] as! [[String: Any]]))
        
        return SPConsentGranted.build { (SPConsentGrantedBuilder) in
            SPConsentGrantedBuilder.setDocumentId(dict?["documentId"] as! String)
            SPConsentGrantedBuilder.setVersion(dict?["documentVersion"] as! String)

            SPConsentGrantedBuilder.setExpiry(dict?["expiry"] as? String)
            SPConsentGrantedBuilder.setName(dict?["documentName"] as? String)
            SPConsentGrantedBuilder.setDescription(dict?["documentDescription"] as? String)

            SPConsentGrantedBuilder.setDocuments(documents)
            SPConsentGrantedBuilder.setContexts(
                getContexts(dict?["contexts"] as? [[String: Any]])
            )
        }
    }
    
    static public func getTiming(dict: [String: Any]?) -> SPTiming {
        return SPTiming.build { (SPTimingBuilder) in
            SPTimingBuilder.setCategory(dict?["category"] as! String)
            SPTimingBuilder.setVariable(dict?["variable"] as! String)
            SPTimingBuilder.setTiming(dict?["timing"] as! Int)

            SPTimingBuilder.setLabel(dict?["label"] as? String)

            SPTimingBuilder.setContexts(
                getContexts(dict?["contexts"] as? [[String: Any]])
            )
        }
    }
    
    static public func getEcommTransaction(dict: [String: Any]?) -> SPEcommerce {
        let items = getEcommTransactionItems(ecommerceItems: dict?["items"] as! [[String: Any]])
        
        return SPEcommerce.build { (SPEcommTransactionBuilder) in
            SPEcommTransactionBuilder.setOrderId(dict?["orderId"] as! String)
            SPEcommTransactionBuilder.setTotalValue(dict?["totalValue"] as! Double)
            SPEcommTransactionBuilder.setAffiliation(dict?["affiliation"] as? String)
            SPEcommTransactionBuilder.setTaxValue(dict?["taxValue"] as! Double)
            SPEcommTransactionBuilder.setShipping(dict?["shipping"] as! Double)
            SPEcommTransactionBuilder.setCity(dict?["city"] as? String)
            SPEcommTransactionBuilder.setState(dict?["state"] as? String)
            SPEcommTransactionBuilder.setCountry(dict?["country"] as? String)
            SPEcommTransactionBuilder.setCurrency(dict?["currency"] as? String)
            SPEcommTransactionBuilder.setItems(items)
            SPEcommTransactionBuilder.setContexts(
                getContexts(dict?["contexts"] as? [[String: Any]])
            )
        }
    }
    
    static public func getPushNotification(dict: [String: Any]?) -> SPPushNotification {
        let notification = getNotificationContent(dict: (dict?["notificationContent"] as? [String: Any])!)
        
        return SPPushNotification.build { (SPPushNotificationBuilder) in
            SPPushNotificationBuilder.setAction((dict?["action"] as? String)!)
            SPPushNotificationBuilder.setDeliveryDate((dict?["deliveryDate"] as? String)!)
            SPPushNotificationBuilder.setTrigger((dict?["trigger"] as? String)!)
            SPPushNotificationBuilder.setCategoryIdentifier((dict?["categoryIdentifier"] as? String)!)
            SPPushNotificationBuilder.setThreadIdentifier((dict?["threadIdentifier"] as? String)!)
            SPPushNotificationBuilder.setNotification(notification)
            SPPushNotificationBuilder.setContexts(
                getContexts(dict?["contexts"] as? [[String: Any]])
            )
        }
    }
    
    static private func getSelfDescribingJson(dict: [String: Any]) -> SPSelfDescribingJson {
        let payload: SPPayload = SPPayload(nsDictionary: dict["payload"] as? [String: NSObject])
        return SPSelfDescribingJson(schema: dict["schema"] as? String, andPayload: payload)
    }
    
    static private func getConsentDocuments(payloads: Array<[String: Any]>) -> Array<SPSelfDescribingJson> {
        return payloads.map({ (element: [String: Any]) -> SPSelfDescribingJson in
            let payload = SPPayload(nsDictionary: element as? [String: NSObject])
            return SPSelfDescribingJson(schema: kSPConsentDocumentSchema, andPayload: payload)
        })
    }
    
    static private func getEcommTransactionItems(ecommerceItems: Array<[String: Any]>) -> Array<SPEcommerceItem> {
        return ecommerceItems.map { (ecommerceItem: [String : Any]) -> SPEcommerceItem in
            return SPEcommerceItem.build { (SPEcommTransactionItemBuilder) in
                (ecommerceItem["itemId"] as? String).map(SPEcommTransactionItemBuilder.setItemId)
                (ecommerceItem["sku"] as? String).map(SPEcommTransactionItemBuilder.setSku)
                (ecommerceItem["price"] as? Double).map(SPEcommTransactionItemBuilder.setPrice)
                (ecommerceItem["quantity"] as? Int).map(SPEcommTransactionItemBuilder.setQuantity)

                SPEcommTransactionItemBuilder.setName(ecommerceItem["name"] as? String)
                SPEcommTransactionItemBuilder.setCategory(ecommerceItem["category"] as? String)
                SPEcommTransactionItemBuilder.setCurrency(ecommerceItem["currency"] as? String)

                SPEcommTransactionItemBuilder.setContexts(
                    getContexts(ecommerceItem["contexts"] as? [[String: Any]])
                )
            }
        }
    }
    
    static private func getNotificationContent(dict: [String: Any]) -> SPNotificationContent {
        return SPNotificationContent.build { (SPNotificationContentBuilder) in
            (dict["badge"] as? Int).map(NSNumber.init).map(SPNotificationContentBuilder.setBadge)
            (dict["title"] as? String).map(SPNotificationContentBuilder.setTitle)
            (dict["body"] as? String).map(SPNotificationContentBuilder.setBody)

            SPNotificationContentBuilder.setSubtitle(dict["subtitle"] as? String)

            SPNotificationContentBuilder.setSound(dict["sounds"] as? String)
            SPNotificationContentBuilder.setLaunchImageName(dict["launchImageName"] as? String)
            SPNotificationContentBuilder.setUserInfo(dict["userInfo"] as? [String: Any])
            SPNotificationContentBuilder.setAttachments(dict["attachments"] as? Array<Any>)
        }
    }

    static private func getContexts(_ contexts: [[String: Any]]?) -> NSMutableArray {
        return NSMutableArray(array: contexts?.map(getSelfDescribingJson) ?? [])
    }
}
